### ISSUE TITLE
[NDTensors] [BUG] `tr` on GPU

### ISSUE THREAD

**Issue by mtfishman** [ITensor/ITensors.jl#1450]
Reported here: https://itensor.discourse.group/t/problem-performing-trace-calculation-with-gpu/1728

```julia
using ITensors: Index, random_itensor
using LinearAlgebra: tr
using Metal: mtl
i = Index(2)
x = random_itensor(i', i)
tr(mtl(x))
```

gives:

<details><summary>output</summary><p>

```julia
Scalar indexing is disallowed.
Invocation of getindex resulted in scalar indexing of a GPU array.
This is typically caused by calling an iterating implementation of a method.
Such implementations *do not* execute on the GPU, but very slowly on the CPU,
and therefore should be avoided.

If you want to allow scalar iteration, use `allowscalar` or `@allowscalar`
to enable scalar iteration globally or for the operations in question.
Stacktrace:
  [1] error(s::String)
    @ Base ./error.jl:35
  [2] errorscalar(op::String)
    @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:155
  [3] _assertscalar(op::String, behavior::GPUArraysCore.ScalarIndexing)
    @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:128
  [4] assertscalar(op::String)
    @ GPUArraysCore ~/.julia/packages/GPUArraysCore/GMsgk/src/GPUArraysCore.jl:116
  [5] getindex
    @ ~/.julia/packages/GPUArrays/OKkAu/src/host/indexing.jl:48 [inlined]
  [6] getindex
    @ ~/.julia/packages/NDTensors/5rDm2/src/lib/Expose/src/functions/abstractarray.jl:13 [inlined]
  [7] getindex
    @ ~/.julia/packages/NDTensors/5rDm2/src/dense/densetensor.jl:90 [inlined]
  [8] getindex
    @ ~/.julia/packages/NDTensors/5rDm2/src/dense/densetensor.jl:95 [inlined]
  [9] getdiagindex
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensor/tensor.jl:361 [inlined]
 [10] contract!(C::NDTensors.DenseTensor{Float32, 0, Tuple{}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, Clabels::Tuple{}, A::NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}, Alabels::Tuple{Int64, Int64}, B::NDTensors.DenseTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, Blabels::Tuple{Int64, Int64}, α::Bool, β::Bool; convert_to_dense::Bool)
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/diag/tensoralgebra/contract.jl:132
 [11] contract!
    @ ~/.julia/packages/NDTensors/5rDm2/src/diag/tensoralgebra/contract.jl:110 [inlined]
 [12] contract!
    @ ~/.julia/packages/NDTensors/5rDm2/src/diag/tensoralgebra/contract.jl:232 [inlined]
 [13] contract! (repeats 2 times)
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:165 [inlined]
 [14] _contract!!
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:143 [inlined]
 [15] _contract!!
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:131 [inlined]
 [16] contract!!
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:219 [inlined]
 [17] contract!!
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:188 [inlined]
 [18] contract(tensor1::NDTensors.DenseTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, labelstensor1::Tuple{Int64, Int64}, tensor2::NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}, labelstensor2::Tuple{Int64, Int64}, labelsoutput_tensor::Tuple{})
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:113
 [19] contract(::Type{NDTensors.CanContract{NDTensors.DenseTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}}}, tensor1::NDTensors.DenseTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, labels_tensor1::Tuple{Int64, Int64}, tensor2::NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}, labels_tensor2::Tuple{Int64, Int64})
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:91
 [20] contract
    @ ~/.julia/packages/SimpleTraits/l1ZsK/src/SimpleTraits.jl:331 [inlined]
 [21] _contract(A::NDTensors.DenseTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, B::NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}})
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/tensor_algebra.jl:3
 [22] _contract(A::ITensors.ITensor, B::ITensors.ITensor)
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/tensor_algebra.jl:9
 [23] contract(A::ITensors.ITensor, B::ITensors.ITensor)
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/tensor_algebra.jl:74
 [24] *(A::ITensors.ITensor, B::ITensors.ITensor)
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/tensor_algebra.jl:61
 [25] _tr(T::ITensors.ITensor; plev::Pair{Int64, Int64}, tags::Pair{ITensors.TagSets.GenericTagSet{BitIntegers.UInt256, 4}, ITensors.TagSets.GenericTagSet{BitIntegers.UInt256, 4}})
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/matrix_algebra.jl:9
 [26] _tr
    @ ~/.julia/packages/ITensors/iifFp/src/tensor_operations/matrix_algebra.jl:2 [inlined]
 [27] #tr#250
    @ ~/.julia/packages/ITensors/iifFp/src/tensor_operations/matrix_algebra.jl:20 [inlined]
 [28] tr(T::ITensors.ITensor)
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/matrix_algebra.jl:19
 [29] top-level scope
    @ REPL[16]:1
 [30] top-level scope
    @ ~/.julia/packages/Metal/q9oGt/src/initialization.jl:57
```

</p></details>

and:

```julia
using ITensorMPS: MPO, siteinds
using LinearAlgebra: tr
using Metal: mtl
s = siteinds(2, 2)
o = MPO(s, "Id")
tr(mtl(o))
```

gives:

<details><summary>output</summary><p>

```julia
ArgumentError: cannot take the CPU address of a Metal.MtlMatrix{Float32, Metal.MTL.Private}
Stacktrace:
  [1] unsafe_convert(::Type{Ptr{Float32}}, x::Metal.MtlMatrix{Float32, Metal.MTL.Private})
    @ Metal ~/.julia/packages/Metal/q9oGt/src/array.jl:197
  [2] gemm!(transA::Char, transB::Char, alpha::Float32, A::Matrix{Float32}, B::Metal.MtlMatrix{Float32, Metal.MTL.Private}, beta::Float32, C::Metal.MtlMatrix{Float32, Metal.MTL.Private})
    @ LinearAlgebra.BLAS ~/.julia/juliaup/julia-1.10.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/blas.jl:1524
  [3] gemm_wrapper!(C::Metal.MtlMatrix{Float32, Metal.MTL.Private}, tA::Char, tB::Char, A::Matrix{Float32}, B::Metal.MtlMatrix{Float32, Metal.MTL.Private}, _add::LinearAlgebra.MulAddMul{true, true, Float32, Float32})
    @ LinearAlgebra ~/.julia/juliaup/julia-1.10.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/matmul.jl:605
  [4] generic_matmatmul!(C::Metal.MtlMatrix{Float32, Metal.MTL.Private}, tA::Char, tB::Char, A::Matrix{Float32}, B::Metal.MtlMatrix{Float32, Metal.MTL.Private}, _add::LinearAlgebra.MulAddMul{true, true, Float32, Float32})
    @ LinearAlgebra ~/.julia/juliaup/julia-1.10.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/matmul.jl:352
  [5] mul!
    @ ~/.julia/juliaup/julia-1.10.3+0.aarch64.apple.darwin14/share/julia/stdlib/v1.10/LinearAlgebra/src/matmul.jl:263 [inlined]
  [6] mul!(CM::NDTensors.Expose.Exposed{Metal.MtlMatrix{Float32, Metal.MTL.Private}, Metal.MtlMatrix{Float32, Metal.MTL.Private}}, AM::NDTensors.Expose.Exposed{Matrix{Float32}, LinearAlgebra.Transpose{Float32, Matrix{Float32}}}, BM::NDTensors.Expose.Exposed{Metal.MtlMatrix{Float32, Metal.MTL.Private}, Metal.MtlMatrix{Float32, Metal.MTL.Private}}, α::Float32, β::Float32)
    @ NDTensors.Expose ~/.julia/packages/NDTensors/5rDm2/src/lib/Expose/src/functions/mul.jl:2
  [7] mul!!(CM::Metal.MtlMatrix{Float32, Metal.MTL.Private}, AM::LinearAlgebra.Transpose{Float32, Matrix{Float32}}, BM::Metal.MtlMatrix{Float32, Metal.MTL.Private}, α::Float32, β::Float32)
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/abstractarray/mul.jl:2
  [8] _contract!(CT::Metal.MtlVector{Float32, Metal.MTL.Private}, AT::Matrix{Float32}, BT::Metal.MtlArray{Float32, 3, Metal.MTL.Private}, props::NDTensors.ContractionProperties{2, 3, 1}, α::Bool, β::Bool)
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/abstractarray/tensoralgebra/contract.jl:174
  [9] _contract!(CT::NDTensors.DenseTensor{Float32, 1, Tuple{Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, AT::NDTensors.DenseTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Vector{Float32}}}, BT::NDTensors.DenseTensor{Float32, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, props::NDTensors.ContractionProperties{2, 3, 1}, α::Bool, β::Bool)
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/dense/tensoralgebra/contract.jl:228
 [10] contract!
    @ ~/.julia/packages/NDTensors/5rDm2/src/dense/tensoralgebra/contract.jl:211 [inlined]
 [11] contract!(C::NDTensors.DenseTensor{Float32, 1, Tuple{Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, Clabels::Tuple{Int64}, A::NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}, Alabels::Tuple{Int64, Int64}, B::NDTensors.DenseTensor{Float32, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, Blabels::Tuple{Int64, Int64, Int64}, α::Bool, β::Bool; convert_to_dense::Bool)
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/diag/tensoralgebra/contract.jl:148
 [12] contract!
    @ ~/.julia/packages/NDTensors/5rDm2/src/diag/tensoralgebra/contract.jl:110 [inlined]
 [13] contract!
    @ ~/.julia/packages/NDTensors/5rDm2/src/diag/tensoralgebra/contract.jl:232 [inlined]
 [14] contract! (repeats 2 times)
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:165 [inlined]
 [15] _contract!!
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:143 [inlined]
 [16] _contract!!
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:131 [inlined]
 [17] contract!!
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:219 [inlined]
 [18] contract!!
    @ ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:188 [inlined]
 [19] contract(tensor1::NDTensors.DenseTensor{Float32, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, labelstensor1::Tuple{Int64, Int64, Int64}, tensor2::NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}, labelstensor2::Tuple{Int64, Int64}, labelsoutput_tensor::Tuple{Int64})
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:113
 [20] contract(::Type{NDTensors.CanContract{NDTensors.DenseTensor{Float32, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}}}, tensor1::NDTensors.DenseTensor{Float32, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, labels_tensor1::Tuple{Int64, Int64, Int64}, tensor2::NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}}, labels_tensor2::Tuple{Int64, Int64})
    @ NDTensors ~/.julia/packages/NDTensors/5rDm2/src/tensoroperations/generic_tensor_operations.jl:91
 [21] contract
    @ ~/.julia/packages/SimpleTraits/l1ZsK/src/SimpleTraits.jl:331 [inlined]
 [22] _contract(A::NDTensors.DenseTensor{Float32, 3, Tuple{Index{Int64}, Index{Int64}, Index{Int64}}, NDTensors.Dense{Float32, Metal.MtlVector{Float32, Metal.MTL.Private}}}, B::NDTensors.DiagTensor{Float32, 2, Tuple{Index{Int64}, Index{Int64}}, NDTensors.Diag{Float32, Float32}})
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/tensor_algebra.jl:3
 [23] _contract(A::ITensors.ITensor, B::ITensors.ITensor)
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/tensor_algebra.jl:9
 [24] contract(A::ITensors.ITensor, B::ITensors.ITensor)
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/tensor_algebra.jl:74
 [25] *(A::ITensors.ITensor, B::ITensors.ITensor)
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/tensor_algebra.jl:61
 [26] _tr(T::ITensors.ITensor; plev::Pair{Int64, Int64}, tags::Pair{ITensors.TagSets.GenericTagSet{BitIntegers.UInt256, 4}, ITensors.TagSets.GenericTagSet{BitIntegers.UInt256, 4}})
    @ ITensors ~/.julia/packages/ITensors/iifFp/src/tensor_operations/matrix_algebra.jl:9
 [27] _tr
    @ ~/.julia/packages/ITensors/iifFp/src/tensor_operations/matrix_algebra.jl:2 [inlined]
 [28] #tr#250
    @ ~/.julia/packages/ITensors/iifFp/src/tensor_operations/matrix_algebra.jl:20 [inlined]
 [29] tr
    @ ~/.julia/packages/ITensors/iifFp/src/tensor_operations/matrix_algebra.jl:19 [inlined]
 [30] tr(M::MPO; plev::Pair{Int64, Int64}, tags::Pair{ITensors.TagSets.GenericTagSet{BitIntegers.UInt256, 4}, ITensors.TagSets.GenericTagSet{BitIntegers.UInt256, 4}})
    @ ITensors.ITensorMPS ~/.julia/packages/ITensors/iifFp/src/lib/ITensorMPS/src/mpo.jl:553
 [31] tr(M::MPO)
    @ ITensors.ITensorMPS ~/.julia/packages/ITensors/iifFp/src/lib/ITensorMPS/src/mpo.jl:542
 [32] top-level scope
    @ REPL[23]:1
 [33] top-level scope
    @ ~/.julia/packages/Metal/q9oGt/src/initialization.jl:57
```

</p></details>

with versions:

```julia
Status `.../Project.toml`
  [0d1a4710] ITensorMPS v0.2.2
  [9136182c] ITensors v0.6.10
  [dde4c033] Metal v1.1.0
```

**Comment by mtfishman** [comment]
@kmp5VT looks like there are a few interconnected issues here stemming from issues contracting `delta` tensors with GPU arrays.

**Comment by mtfishman** [comment]
Would also be good to make sure that whatever changes are made to fix this also fix these operations when QN conservation/block sparse tensors are used, those probably have their own set of challenges to get them working on GPU.

**Comment by mtfishman** [comment]
Closed by #1453.
